### ISSUE TITLE
Add cache warmer

### ISSUE THREAD

**Issue by jaikdean** [willdurand/BazingaHateoasBundle#80]
It would be really useful to have a cache warmer registered automatically. We've done this manually by adding the following to our `services.yaml`:

```yaml
    hateoas.configuration.metadata.cache.cache_warmer:
        class: JMS\SerializerBundle\Cache\CacheWarmer
        arguments:
            $includePaths: ['%kernel.project_dir%/src/Entity']
            $metadataFactory: '@hateoas.configuration.metadata_factory'
        tags: [kernel.cache_warmer]
```

**Comment by tonivdv** [comment]
@jaikdean Do you confirm that if you really want your warmup to be 100% done you will also have to include the files inside the hateoas library? Otherwise in production, if cache is only readable it will fail because e.g. Hateoas-Representation-CollectionRepresentation.cache.php will not be generated.

**Comment by jaikdean** [comment]
@tonivdv Out of the box with Symfony, that quote from the documentation is true. However, third-party bundles might try to write there regardless of the env/debug settings.

I'm not sure about the issue you're having I'm afraid, I've not touched this since I opened the original issue, and the solution in my first post seems to work for us. We're probably only using a subset of the functionality in this bundle though.

**Comment by tonivdv** [comment]
@jaikdean Ok. And also it depends if you made the /var/cache writable or not ... if writable then you don't "detect" this kind of issue

@goetas What do you mean with that? I currently took the snippet from @jaikdean which does warm the cache up, but still could the above error. Unless you mean something else with the reference?

Thanks

**Comment by goetas** [comment]
@jaikdean is actually using the feature implemented in schmittjoh/JMSSerializerBundle#615

**Comment by tonivdv** [comment]
@goetas Yes I understood that. But by doing like @jaikdean is doing:

```yml
hateoas.configuration.metadata.cache.cache_warmer:
        class: JMS\SerializerBundle\Cache\CacheWarmer
        arguments:
            $includePaths: ['%kernel.project_dir%/src/Entity']
            $metadataFactory: '@hateoas.configuration.metadata_factory'
        tags: [kernel.cache_warmer]
```

It does not take into account some classes present in the hateoas library. So I changed to:

```yml
# Cache Warmer Service for Hateoas. Necessary until:
    hateoas.configuration.metadata.cache.cache_warmer:
        class: JMS\SerializerBundle\Cache\CacheWarmer
        arguments:
          $includePaths: ['%kernel.project_dir%/src', '%kernel.project_dir%/vendor/willdurand/hateoas/src/Hateoas/Representation']
          $metadataFactory: '@hateoas.configuration.metadata_factory'
        tags: [kernel.cache_warmer]
```

But this still didn't do the trick while it did add the files in the var/cache/prod/hateoas directory.

I'm playing with this at the moment and what I just discovered is that it also needs to have those files in src/Hateoas/Representation for the JMS. 

In conclusion I need to tell JMS warmup to also take into account src/Hateoas/Representation.

I will try that and let you guys know if it worked ... If it does ... wouldn't this be nice to have this configured in this bundle as per @jaikdean 's suggestion?

**Comment by tonivdv** [comment]
Ok so that was the issue. The next configuration solves the issue and allows you to keep you cache folder read-only in production (assuming you don't have any other third party hitting the cache directory as mentioned by @jaikdean ):

**config/jms_serializer.yaml**
```yml
jms_serializer:

  metadata:
    warmup:
      paths:
        included:
          - '%kernel.root_dir%'
          - '%kernel.project_dir%/vendor/willdurand/hateoas/src/Hateoas/Representation'
```
What's important in the above snippet is the second include. The first one can be enhanced if you're files are more regrouped.

config/services.yaml
```yml
hateoas.configuration.metadata.cache.cache_warmer:
    class: JMS\SerializerBundle\Cache\CacheWarmer
    arguments:
      $includePaths: ['%kernel.project_dir%/src', '%kernel.project_dir%/vendor/willdurand/hateoas/src/Hateoas/Representation']
      $metadataFactory: '@hateoas.configuration.metadata_factory'
    tags: [kernel.cache_warmer]
```

@goetas Wouldn't it be useful to have a part of this config in the bundle itself? Or at least have it documented?!

**Comment by goetas** [comment]
I agree that would be to have it. A pr (with tests) would be welcome!
